pest/src/parser_state.rs

```diff
@@ -50,6 +53,49 @@ pub enum MatchDir {
     TopToBottom,
 }
 
+static CALL_LIMIT: AtomicUsize = AtomicUsize::new(0);
+
+/// Sets the maximum call limit for the parser state
+/// to prevent stack overflows or excessive execution times
+/// in some grammars.
+/// If set, the calls are tracked as a running total
+/// over all non-terminal rules that can nest closures
+/// (which are passed to transform the parser state).
+///
+/// # Arguments
+///
+/// * `limit` - The maximum number of calls. If None,
+///             the number of calls is unlimited.
+pub fn set_call_limit(limit: Option<NonZeroUsize>) {
+    CALL_LIMIT.store(limit.map(|f| f.get()).unwrap_or(0), Ordering::Relaxed);
+}
+
+#[derive(Debug)]
+struct CallLimitTracker {
+    current_call_limit: Option<(usize, usize)>,
+}
+
+impl Default for CallLimitTracker {
+    fn default() -> Self {
+        let limit = CALL_LIMIT.load(Ordering::Relaxed);
+        let current_call_limit = if limit > 0 { Some((0, limit)) } else { None };
+        Self { current_call_limit }
+    }
+}
+
+impl CallLimitTracker {
+    fn limit_reached(&self) -> bool {
+        self.current_call_limit
+            .map_or(false, |(current, limit)| current >= limit)
+    }
+
+    fn increment_depth(&mut self) {
+        if let Some((current, _)) = &mut self.current_call_limit {
+            *current += 1;
+        }
```

Comment:
@tomtau @NoahTheDuke 
Just for interest, there's yet another way to code this:
```
        if let Some((ref mut current, _)) = self.current_call_limit {
            *current += 1;
        }
```
This cropped up [here](https://youtu.be/rAl-9HwD858?t=2752) in @jonhoo's video [Crust of Rust: Lifetime Annotations](https://www.youtube.com/watch?v=rAl-9HwD858).

Reply:
Oh that's cool. I didn't know you could embed those.